### PR TITLE
bugfix: masks and biases can now be passed with O(n) shapes

### PR DESCRIPTION
**Pull request by xloem**
Further unimplemented ideas include:
- a callback that provides these
- masks or biases that broadcast across chunks

### CODE REVIEW

**Comment by AminRezaei0x443** [comment]
Thanks for your contribution on this,
From what I understood, you're applying the same bias and mask across chunks. The point I can't get is the actual use case of this scenario. For example, one of the typical masks is the casual language model mask which zeros out the future dependency links in sentences. Could you please provide incorporating implemented code to some use cases like that?
A Testcase to do this would be great.

**Comment by xloem** [comment]
thanks for your reply :) i drafted that earlier with thorough testing at https://github.com/xloem/memory-efficient-attention/tree/callbacks but i need to fix the dimension errors my psychosis left in the code, which could take me some time.

i submitted this because i realised that it was impossible to use the masks and weights efficiently in this repository with jax. input attention masks have the requested property.

i bumped into the causal mask afterward. do you think it is useful? existing pretrained causal models take such short inputs. i haven't explored whether they can be finetuned with increased position embeddings. [maybe i should run a test showing this code can make gpt-j process an entire paper or book, to inspire my psychotic parts to help out more.]

anyway what you ask sounds great.

**Comment by xloem** [comment]
to clarify, the "perceiver" architecture of masked language model uses a mask like this, it's where i started prototyping.

**Comment by xloem** [comment]
in this branch i have modified the huggingface perceiver model to use your library: https://github.com/xloem/transformers/tree/memory-efficient-attention

i have some code where a model is trained and used and can find it if you would like. i think i will focus on handling my issues around the callbacks approach, though. not sure what the beset solution here or if problems will be encountered when jit() happens. not very familiar with jax.lax myself.

**Comment by AminRezaei0x443** [comment]
Thanks really for your contribution, unfortunately, I'm a little busy and can't review really accurately right now.
Do you think those `if`s are safe with any input? I mean any possible collision like picking the wrong `if` route.
I think a more readable alternative would be getting functions as input like `mask_calc_fn` and `bias_calc_fn`. Then refactor the main function to decide behavior with its inputs. I hope Jax's stack won't make an issue here.
Anyway, If you think this PR is good, just do a favor and make the `data` method in tests return a list of tuples instead of just one, so it would be tested with both data, And I'll gladly merge your PR.

**Comment by xloem** [comment]
@AminRezaei0x443 sorry for the previous comment spam. how does this look? i tried to do something for everything you mentioned. I kept the mask/bias broadcasting conditions in but flattened them; unsure if removal is better.

**Review by AminRezaei0x443 — APPROVED** [review]
It looks great, Thank you for the contribution! I'm merging the changes.
